version: '3'

networks:
  my-network:
    driver: bridge

services:
  svc-account:
    build: 
      context: ./account
      # dockerfile: ./Dockerfile
    depends_on:
      - database
    environment:
      - DB_HOST=database
      - DB_NAME=accounts
      - API_PORT=8000
    networks:
      - my-network
    ports:
      - 8000:8000

  svc-poller:
    build: 
      context: ./poller
      # dockerfile: ./Dockerfile
    depends_on:
      - database
      - pulsar
    environment:
      - DB_HOST=database
      - DB_NAME=accounts
      - PULSAR_ADDRESS=pulsar:6650
      - TOPIC=accounts
    networks:
      - my-network

  svc-balance:
    build: 
      context: ./balance
      # dockerfile: ./Dockerfile
    depends_on:
      - pulsar
    environment:
      - PULSAR_ADDRESS=pulsar:6650
      - TOPIC=accounts
    networks:
      - my-network

  database:
    image: postgres:12.3-alpine
    healthcheck:
      test: [ "CMD", "pg_isready", "-q", "-d", "postgres", "-U", "root" ]
      timeout: 45s
      interval: 10s
      retries: 10
    restart: always
    environment:
      - POSTGRES_USER=root
      - POSTGRES_PASSWORD=password
      - POSTGRES_DB=accounts
    volumes:
      - ./scripts/db:/docker-entrypoint-initdb.d/
    networks:
      - my-network
    ports:
      - 5432:5432

  pulsar:
    image: apachepulsar/pulsar:2.6.0
    volumes:
      - pulsardata:/pulsar/data
      - pulsarconf:/pulsar/conf
    networks:
      - my-network
    ports:
      - 6650:6650
      - 8080:8080
    command: bin/pulsar standalone
volumes:
  pulsardata:
  pulsarconf: